Scheduled Updates: Limit plugins number up to ten per schedule

ouikhuan opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Fixes #36369

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* Prevent users from creating a schedule with more than 10 plugins. We've already limited this in the FE, and to unify both FE and BE, we are adding a check in this PR to handle this use case. 

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Do a POST request to `/wpcom/v2/update-schedules` with more than 10 plugins, and you'll receive an error code with `rest_invalid_param` and an error message showing "You can schedule a maximum of 10 plugins at one time."
* Run unit test and make sure there's no fails.

ouikhuan committed:
Add test case to cover over 10 plugins on creating a schedule

ouikhuan committed:
Only allow users to create a schedule within 10 plugins

ouikhuan committed:
changelog

ouikhuan committed:
Fix version mismatch

obenland commented on projects/packages/scheduled-updates/src/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-update-schedules.php:
```diff
@@ -328,6 +328,14 @@ public function delete_item( $request ) {
+				__( 'You can schedule a maximum of 10 plugins at one time.', 'jetpack-scheduled-updates' ),
+				array( 'status' => 400 )
+			);
+		}
```
The REST API supports a `maxItems` argument in the endpoint's args definition, which we should be able to use here?
https://github.com/WordPress/WordPress/blob/44c2d5685a161b9a7b52a8f743e7fb96b2cf5666/wp-includes/rest-api.php#L2464-L2478

ouikhuan replied:
Hmmm for some reason it's not working after setting `maxItems`. I don't see any similar usage in `jetpack` repo either, only on wpcom codebase. Will take a closer look at this tomorrow.

obenland replied:
It looks like when a validation callback is provided the built-in schema validation gets skipped. https://github.com/Automattic/jetpack/pull/36375/commits/89f01c0b92f420abd97123d4823247e0dc9d948a brings that back "manually". 

What's odd is that I tried to pass a 'pattern' parameter in the `items` array, [like the wp/v2/plugins endpoint does](https://github.com/WordPress/WordPress/blob/master/wp-includes/rest-api/endpoints/class-wp-rest-plugins-controller.php#L94), but I was not able to make it fail. Maybe I'm missing something.

ouikhuan replied:
> It looks like when a validation callback is provided the built-in schema validation gets skipped. https://github.com/Automattic/jetpack/commit/89f01c0b92f420abd97123d4823247e0dc9d948a brings that back "manually".

Thanks for jumping in and making the changes!

> What's odd is that I tried to pass a 'pattern' parameter in the items array, [like the wp/v2/plugins endpoint does](https://github.com/WordPress/WordPress/blob/master/wp-includes/rest-api/endpoints/class-wp-rest-plugins-controller.php#L94), but I was not able to make it fail. Maybe I'm missing something.

I was doing some testing and noticed something interesting. When I put the `pattern` parameter inside the `items` array, it actually works. But, it looks like the regex in `self::PATTERN` might be the culprit behind our little issue. Just out of curiosity, what examples did you try to catch a fail? I gave `.plugin` and `plugin..file` a shot, and surprisingly, both sneaked past our regex check. It does fail when I try using `./`.

obenland replied:
I used the same pattern as the plugins endpoint, but maybe that's not needed at all.

obenland committed:
Use rest function to validate plugins from schema

When a validation callback is provided the built-in schema validation gets skipped. This will make up for it and we can extend it in the future if needed.
